### PR TITLE
WIP ARROW-2193: [C++] Do not depend on Boost libraries at runtime in plasma_store

### PR DESCRIPTION
**Pull request by wesm**
This is sort of a hack; I wasn't sure the way to deal with this more generally. Unfortunately, this only gets rid of the boost_system and boost_filesystem runtime dependencies. boost_regex still has a transitive dependency somehow

### CODE REVIEW

**Comment by pitrou** [comment]
If this PR takes some time to polish, perhaps it would be nice in the meantime to force LD_LIBRARY_PATH in `test_plasma.py`. What do you think?

File: cpp/src/plasma/CMakeLists.txt [hunk 1]
```diff
@@ -123,6 +123,18 @@ endif()
 
 add_executable(plasma_store store.cc)
 target_link_libraries(plasma_store plasma_static ${PLASMA_LINK_LIBS})
+SET_TARGET_PROPERTIES(plasma_store PROPERTIES
+  LINK_FLAGS "-Wl,--as-needed")
```
Comment (xhochy) [inline cpp/src/plasma/CMakeLists.txt]: Can you use https://cmake.org/cmake/help/v3.0/module/CheckCXXCompilerFlag.html and add this flag globally? If the compiler supports it, we should use it everywhere.

File: cpp/src/plasma/CMakeLists.txt [hunk 2]
```diff
@@ -123,6 +123,18 @@ endif()
...
+
+if (ARROW_RPATH_ORIGIN)
```
Comment (robertnishihara) [inline cpp/src/plasma/CMakeLists.txt]: I think this block could benefit from a short comment explaining what the purpose is.

Comment (wesm) [reply cpp/src/plasma/CMakeLists.txt]: This looks like a rebase artifact

Comment (xhochy) [reply cpp/src/plasma/CMakeLists.txt]: Yes, I have added this in another commit (without it, the manylinux1 wheel would fail).

**Comment by wesm** [comment]
I am not sure why libboost_regex is still a runtime dependency, even with `--as-needed` -- Plasma doesn't appear to have any symbols with a transitive dependency on code in arrow/util/decimal.cc (where boost::regex is used)

**Comment by wesm** [comment]
@xhochy It seems that `CHECK_CXX_COMPILER_FLAG` doesn't turn up the right answer on macOS. I'm afraid we'll have to leave this PR in WIP unless someone else can figure this out for 0.9.0

**Comment by pitrou** [comment]
Should probably close this PR as this issue has been fixed by removing regex_boost usage.

**Pull request closed without merging**